irohad/ordering/impl/ordering_gate_cache/on_demand_cache.cpp

```diff
@@ -13,24 +15,47 @@ using namespace iroha::ordering::cache;
 // TODO: IR-1864 13.11.18 kamilsa use nvi to separate business logic and locking
 // logic
 
-void OnDemandCache::addToBack(
+namespace {
+  uint64_t countTransactions(const OrderingGateCache::BatchesSetType &batches) {
+    return std::accumulate(batches.begin(),
+                           batches.end(),
+                           0u,
+                           [](const uint64_t txs_quantity, const auto &batch) {
+                             return txs_quantity
+                                 + boost::size(batch->transactions());
+                           });
+  }
+}  // namespace
+
+OnDemandCache::OnDemandCache(uint64_t max_cache_size)
+    : max_cache_size_(max_cache_size) {}
+
+bool OnDemandCache::addToBack(
     const OrderingGateCache::BatchesSetType &batches) {
   std::unique_lock<std::shared_timed_mutex> lock(mutex_);
-  circ_buffer.back().insert(batches.begin(), batches.end());
+  const uint64_t input_size = countTransactions(batches);
+
+  if (circ_buffer.back().first + input_size > max_cache_size_) {
```

Comment:
as you assume there is only one batch in the collection, maybe we should make the interface to explicitly accept a single batch?